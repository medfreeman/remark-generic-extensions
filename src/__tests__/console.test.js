import sinon from "sinon";

import { transformToHtml } from "./helpers/transform";

describe("console", () => {
  let originalConsoleError;

  const setup = () => {
    originalConsoleError = console.error;
    console.error = sinon.spy();
  };

  const cleanup = () => {
    console.error = originalConsoleError;
  };

  beforeEach(setup);

  afterEach(cleanup);

  test("should not log an error message to console by default", () => {
    transformToHtml("!icon");
    expect(console.error.callCount).toBe(0);
  });

  test("should log a console error message when `replace` option is not a function", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          // eslint-disable-next-line no-unused-vars
          replace: {}
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test("should log a console error message when `html` and `replace` options are provided", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          html: {},
          // eslint-disable-next-line no-unused-vars
          replace: function(a, b) {}
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test("should log a console error message when an invalid schema is provided", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          html: {
            tagName: "div",
            children: [
              {
                type: "grooow"
              }
            ]
          }
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test("should log a console error message when a comment node hasn't a value", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          html: {
            tagName: "div",
            children: [
              {
                type: "comment"
              }
            ]
          }
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test("should log a console error message when a text node hasn't a value", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          html: {
            tagName: "div",
            children: [
              {
                type: "text"
              }
            ]
          }
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test("should log a console error message when an element node hasn't a tagName", () => {
    transformToHtml("!icon", {
      elements: {
        icon: {
          html: {
            tagName: "div",
            children: [
              {
                type: "element"
              }
            ]
          }
        }
      }
    });

    expect(console.error.callCount).toBe(1);
    for (let index = 0; index < console.error.callCount; index++) {
      const args = console.error.getCall(index).args;
      expect(args.toString()).toMatchSnapshot();
    }
  });

  test(
    "should log a console error message " +
      "when an element node has a dashed property",
    () => {
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              tagName: "div",
              children: [
                {
                  type: "element",
                  tagName: "div",
                  properties: {
                    "t-a": "huu"
                  }
                }
              ]
            }
          }
        }
      });

      expect(console.error.callCount).toBe(1);
      for (let index = 0; index < console.error.callCount; index++) {
        const args = console.error.getCall(index).args;
        expect(args.toString()).toMatchSnapshot();
      }
    }
  );

  test(
    "should log a console error message " +
      "when the top-level node has a forbidden property",
    () => {
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              tagName: "div",
              properties: {
                class: "foo"
              }
            }
          }
        }
      });

      expect(console.error.callCount).toBe(1);
      for (let index = 0; index < console.error.callCount; index++) {
        const args = console.error.getCall(index).args;
        expect(args.toString()).toMatchSnapshot();
      }
    }
  );

  test(
    "should log a console error message " +
      "when a children node has a forbidden property",
    () => {
      transformToHtml("!icon", {
        elements: {
          icon: {
            html: {
              tagName: "div",
              children: [
                {
                  type: "element",
                  tagName: "div",
                  properties: {
                    class: "foo"
                  }
                }
              ]
            }
          }
        }
      });

      expect(console.error.callCount).toBe(1);
      for (let index = 0; index < console.error.callCount; index++) {
        const args = console.error.getCall(index).args;
        expect(args.toString()).toMatchSnapshot();
      }
    }
  );
});
