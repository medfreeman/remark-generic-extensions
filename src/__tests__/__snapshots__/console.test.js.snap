// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`console should log a console error message when a children node has a forbidden property 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"element\\",
            \\"tagName\\": \\"div\\",
            \\"properties\\": {
              \\"class\\": \\"foo\\"
            }
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"class\\\\\\" is not allowed\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"properties\\",
          \\"class\\"
        ],
        \\"type\\": \\"any.unknown\\",
        \\"context\\": {
          \\"key\\": \\"class\\",
          \\"label\\": \\"class\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"element\\",
                \\"tagName\\": \\"div\\",
                \\"properties\\": {
                  \\"class\\": \\"foo\\"
                }
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"element\\",
              \\"tagName\\": \\"div\\",
              \\"properties\\": {
                \\"class\\": \\"foo\\"
              }
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when a comment node hasn't a value 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"comment\\"
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"value\\\\\\" is required\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"value\\"
        ],
        \\"type\\": \\"any.required\\",
        \\"context\\": {
          \\"key\\": \\"value\\",
          \\"label\\": \\"value\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"comment\\"
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"comment\\"
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when a text node hasn't a value 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"text\\"
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"value\\\\\\" is required\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"value\\"
        ],
        \\"type\\": \\"any.required\\",
        \\"context\\": {
          \\"key\\": \\"value\\",
          \\"label\\": \\"value\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"text\\"
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"text\\"
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when an element node has a dashed property 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"element\\",
            \\"tagName\\": \\"div\\",
            \\"properties\\": {
              \\"t-a\\": \\"huu\\"
            }
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"t-a\\\\\\" is not allowed\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"properties\\",
          \\"t-a\\"
        ],
        \\"type\\": \\"object.allowUnknown\\",
        \\"context\\": {
          \\"child\\": \\"t-a\\",
          \\"key\\": \\"t-a\\",
          \\"label\\": \\"t-a\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"element\\",
                \\"tagName\\": \\"div\\",
                \\"properties\\": {
                  \\"t-a\\": \\"huu\\"
                }
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"element\\",
              \\"tagName\\": \\"div\\",
              \\"properties\\": {
                \\"t-a\\": \\"huu\\"
              }
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when an element node hasn't a tagName 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"element\\"
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"tagName\\\\\\" is required\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"tagName\\"
        ],
        \\"type\\": \\"any.required\\",
        \\"context\\": {
          \\"key\\": \\"tagName\\",
          \\"label\\": \\"tagName\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"element\\"
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"element\\"
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when an invalid schema is provided 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"children\\": [
          {
            \\"type\\": \\"grooow\\"
          }
        ]
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"type\\\\\\" with value \\\\\\"grooow\\\\\\" fails to match the hast node type pattern\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"children\\",
          0,
          \\"type\\"
        ],
        \\"type\\": \\"string.regex.name\\",
        \\"context\\": {
          \\"name\\": \\"hast node type\\",
          \\"pattern\\": {},
          \\"value\\": \\"grooow\\",
          \\"key\\": \\"type\\",
          \\"label\\": \\"type\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"children\\": [
              {
                \\"type\\": \\"grooow\\"
              }
            ]
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"children\\": [
            {
              \\"type\\": \\"grooow\\"
            }
          ]
        }
      }
    }
  }
}"
`;

exports[`console should log a console error message when the top-level node has a forbidden property 1`] = `
"Invalid options provided:
{
  \\"elements\\": {
    \\"icon\\": {
      \\"html\\": {
        \\"tagName\\": \\"div\\",
        \\"properties\\": {
          \\"class\\": \\"foo\\"
        }
      }
    }
  }
}

returned:
{
  \\"error\\": {
    \\"isJoi\\": true,
    \\"name\\": \\"ValidationError\\",
    \\"details\\": [
      {
        \\"message\\": \\"\\\\\\"class\\\\\\" is not allowed\\",
        \\"path\\": [
          \\"elements\\",
          \\"icon\\",
          \\"html\\",
          \\"properties\\",
          \\"class\\"
        ],
        \\"type\\": \\"any.unknown\\",
        \\"context\\": {
          \\"key\\": \\"class\\",
          \\"label\\": \\"class\\"
        }
      }
    ],
    \\"_object\\": {
      \\"elements\\": {
        \\"icon\\": {
          \\"html\\": {
            \\"tagName\\": \\"div\\",
            \\"properties\\": {
              \\"class\\": \\"foo\\"
            }
          }
        }
      }
    }
  },
  \\"value\\": {
    \\"elements\\": {
      \\"icon\\": {
        \\"html\\": {
          \\"tagName\\": \\"div\\",
          \\"properties\\": {
            \\"class\\": \\"foo\\"
          }
        }
      }
    }
  }
}"
`;
